### PR TITLE
[usm] Create entry in `connection_protocol` only if needed

### PR DESCRIPTION
**Pull request by guyarb**
### What does this PR do?

Fixes the current codebase to generate an entry in the `connection_protocol` map only when a connection was classified.

### Motivation

* The map has high utilization rate in production clusters. Once the map is full, we cannot insert new values, and then USM can suffer from inaccuracies when decoding a protocol, and NPM can report wrong protocol for the connection. 
* Reduce pressure on the `connection_protocol` map, but creating an entry only when needed (= when a connection is classified and supported).
* Don't generate entries on `tls_finish` callback.

#### UTs
Added a new test to ensure an unclassified connection does not create an entry in the map.
The test passes on the branch but fails on `main` branch prior to the change.
Test `TestUSMSuite/.*/TestProtocolClassification/.*/unclassified`

Other classification tests are passing.

#### Dogfooding

Ran on `oddish-c` without impacting the accuracy or yielding any error.

### CODE REVIEW

**Review by amitslavin — CHANGES_REQUESTED** [review]

File: pkg/network/usm/tests/tracer_usm_linux_test.go [hunk 1]
```diff
@@ -700,6 +703,91 @@ func TestFullMonitorWithTracer(t *testing.T) {
...
+				targetAddress: targetAddress,
+				extras:        make(map[string]interface{}),
+			},
+			postTracerSetup: func(t *testing.T, ctx testContext) {
+				// Initiate a connection to the TCP server.
+				c, err := net.Dial("tcp", ctx.targetAddress)
+				require.NoError(t, err)
+				ctx.extras["client"] = c
+
+				// Writing a magic and the settings in the same packet to socket.
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/usm/tests/tracer_usm_linux_test.go]: removed

File: pkg/network/usm/tests/tracer_usm_linux_test.go [hunk 2]
```diff
@@ -700,6 +703,91 @@ func TestFullMonitorWithTracer(t *testing.T) {
...
+		{
+			name: "unsupported TCP protocol",
+			context: testContext{
+				serverPort:    rawTrafficPort,
+				serverAddress: serverAddress,
+				targetAddress: targetAddress,
+				extras:        make(map[string]interface{}),
+			},
+			postTracerSetup: func(t *testing.T, ctx testContext) {
+				// Initiate a connection to the TCP server.
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/usm/tests/tracer_usm_linux_test.go]: removed

File: pkg/network/usm/tests/tracer_usm_linux_test.go [hunk 3]
```diff
@@ -700,6 +703,91 @@ func TestFullMonitorWithTracer(t *testing.T) {
 	require.NoError(t, tr.RegisterClient(clientID))
 }
 
+func testUnclassifiedProtocol(t *testing.T, tr *tracer.Tracer, clientHost, targetHost, serverHost string) {
+	// http2 server init
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/usm/tests/tracer_usm_linux_test.go]: removed

File: pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h [hunk 4]
```diff
@@ -157,6 +148,16 @@ static __always_inline void protocol_dispatcher_entrypoint(struct __sk_buff *skb
         log_debug("[protocol_dispatcher_entrypoint]: %p Classifying protocol as: %d", skb, cur_fragment_protocol);
         // If there has been a change in the classification, save the new protocol.
         if (cur_fragment_protocol != PROTOCOL_UNKNOWN) {
+            stack = get_or_create_protocol_stack(&skb_tup);
+            if (!stack) {
+                // should never happen, but it is required by the eBPF verifier
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h:155]: yes, as the comment mentioned, it is header for the verifier to be happy

File: pkg/network/ebpf/c/protocols/classification/shared-tracer-maps.h [hunk 5]
```diff
@@ -15,34 +15,43 @@ static __always_inline bool is_protocol_classification_supported() {
...
+// If the tuple is not found, returns NULL.
+static __always_inline protocol_stack_t* get_protocol_stack_if_exists(conn_tuple_t* tuple) {
+    conn_tuple_t normalized_tup = *tuple;
+    normalize_tuple(&normalized_tup);
+    return __get_protocol_stack_if_exists(&normalized_tup);
+}
+
+// Returns the protocol_stack_t associated with the given connection tuple.
+// If the tuple is not found, creates a new entry and returns it.
+static __always_inline protocol_stack_t* get_or_create_protocol_stack(conn_tuple_t *skb_tup) {
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/classification/shared-tracer-maps.h:37]: `get_protocol_stack_if_exists` returns the stack if it exists in the map
`get_or_create_protocol_stack` returns the stack if if exists, and if not, we add it to the map and then return it

File: pkg/network/ebpf/c/protocols/classification/shared-tracer-maps.h [hunk 6]
```diff
@@ -37,10 +37,9 @@ static __always_inline protocol_stack_t* get_protocol_stack_if_exists(conn_tuple
 static __always_inline protocol_stack_t* get_or_create_protocol_stack(conn_tuple_t *skb_tup) {
     conn_tuple_t normalized_tup = *skb_tup;
     normalize_tuple(&normalized_tup);
-    protocol_stack_wrapper_t* wrapper = bpf_map_lookup_elem(&connection_protocol, &normalized_tup);
+    protocol_stack_t *wrapper = __get_protocol_stack_if_exists(&normalized_tup);
     if (wrapper) {
-        wrapper->updated = bpf_ktime_get_ns();
-        return &wrapper->stack;
+        return wrapper;
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/classification/shared-tracer-maps.h:42]: reason for what?

**Review by amitslavin — APPROVED** [review]

**Review by hmahmood — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/classification/routing.h [hunk 7]
```diff
@@ -55,6 +55,11 @@ static __always_inline void init_routing_cache(usm_context_t *usm_ctx, protocol_
     usm_ctx->routing_skip_layers = 0;
     usm_ctx->routing_current_program = CLASSIFICATION_PROG_UNKNOWN;
 
+    // No protocol stack, nothing to mark for skipping
+    if (!stack) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/classification/routing.h:61]: Before Adam's change, `has_available_program(__PROG_ENCRYPTION)` returned false, now it returns true

File: pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h [hunk 8]
```diff
@@ -124,16 +124,7 @@ static __always_inline void protocol_dispatcher_entrypoint(struct __sk_buff *skb
...
-    protocol_stack_t *stack = get_protocol_stack(&skb_tup);
-    if (!stack) {
-        // should never happen, but it is required by the eBPF verifier
-        return;
-    }
-
-    // This is used to signal the tracer program that this protocol stack
-    // is also shared with our USM program for the purposes of deletion.
-    // For more context refer to the comments in `delete_protocol_stack`
-    stack->flags |= FLAG_USM_ENABLED;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/dispatcher-helpers.h:136]: maybe we should have kept those lines with the simple if clause of
```c
if (stack) {
    // This is used to signal the tracer program that this protocol stack
    // is also shared with our USM program for the purposes of deletion.
    // For more context refer to the comments in `delete_protocol_stack`
    stack->flags |= FLAG_USM_ENABLED;
}
```

**Pull request merged**